src/persist/src/indexed/encoding.rs

```diff
@@ -230,7 +235,36 @@ impl BlobTraceUpdates {
                 // Recurse at most once, since this data is now structured.
                 self.get_or_make_structured::<K, V>(key_schema, val_schema)
             }
-            BlobTraceUpdates::Both(_, structured) => structured,
+            BlobTraceUpdates::Both(_, structured) => {
+                // If the types don't match, attempt to migrate the array to the new type.
+                // We expect this to succeed, since this should only be called with backwards-
+                // compatible schemas... but if it fails we only log, and let some higher-level
+                // code signal the error if it cares.
+                let migrate = |array: &mut ArrayRef, to_type: DataType| {
+                    let from_type = array.data_type().clone();
+                    if from_type != to_type {
```

Comment:
This adds a second way of doing this logic, on top of SchemaIds. Let's add this to [our discussion of 29283](https://github.com/MaterializeInc/materialize/pull/29283#issuecomment-2327492133)

Reply:
When I typed up 29283, I was picturing something like a Uuid as the key for schemas that we don't have a SchemaId for, but I like your suggestion of using the DataType much better. That makes this bit much more aligned, so I'm happy.

I do think we might want to plumb 29283 here (recomputing things feels a bit less bad in compaction, but still may as well use the cache if we have it), but I can rebase my PR on top of this and do that.

Reply:
Cool! FWIW this path is only hit when the structured-compaction flag is on, and that's off even in CI at the moment, so it should be okay to take that as a followup and avoid dependencies between the branches.